src/test/java/org/reaktivity/nukleus/mqtt/internal/ValidatorTest.java

```diff
@@ -0,0 +1,305 @@
+    {
+        final String topicFilter = "##";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+    @Test
+    public void shouldNotValidateAdjacentMultiLevelWildcardWithTrailingLevelSeparator()
+    {
+        final String topicFilter = "##/";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+    @Test
+    public void shouldNotValidateAdjacentMultiLevelWildcardWithLeadingLevelSeparator()
+    {
+        final String topicFilter = "/##";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+    @Test
+    public void shouldNotValidateAdjacentMultiAndSingleLevelWildcards()
+    {
+        final String topicFilter = "#+";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+    @Test
+    public void shouldNotValidateAdjacentSingleLevelWildcards()
+    {
+        final String topicFilter = "++";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+    @Test
+    public void shouldNotValidateAdjacentSingleLevelWildcardsWithTrailingLevelSeparator()
+    {
+        final String topicFilter = "++/";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+    @Test
+    public void shouldNotValidateAdjacentSingleLevelWildcardsWithLeadingLevelSeparator()
+    {
+        final String topicFilter = "/++";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+    @Test
+    public void shouldNotValidateLeadingMultiLevelWildcardsWithMultipleTopicNames()
+    {
+        final String topicFilter = "#/topic/name";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+    @Test
+    public void shouldNotValidateSingleLevelWildcardsCombinedWithTopicName()
+    {
+        final String topicFilter = "+topic/name";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+    @Test
+    public void shouldNotValidateMultiLevelWildcardsCombinedWithTopicName()
+    {
+        final String topicFilter = "#topic/name";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+    @Test
+    public void shouldNotValidateMultiLevelWildcardsCombinedWithTopicNameWithValidSingleLevelWildcard()
+    {
+        final String topicFilter = "#topic/+/name";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+
+    @Test
+    public void shouldNotValidateSingleLevelWildcardsCombinedWithTopicNameWithValidMultiLevelWildcard()
+    {
+        final String topicFilter = "+topic/name/#";
+
+        checkForInvalidTopicFilter(topicFilter);
+    }
+
+    private void checkForValidTopicFilter(
+        String topicFilter)
+    {
+        assertTrue(String.format("\"%s\" was an invalid filter when it's supposed to be valid.", topicFilter),
+            validator.isTopicFilterValid(topicFilter));
+    }
+
+    private void checkForInvalidTopicFilter(
+        String topicFilter)
+    {
+        assertFalse(String.format("\"%s\" was a valid filter when it's supposed to be invalid.", topicFilter),
+            validator.isTopicFilterValid(topicFilter));
+    }
```

Comment:
Now that each test name indicates expectations directly, no need for descriptive assertion failure message as well.

Please inline all calls to `checkForValidTopicFilter` and `checkForValidTopicFilter` and then remove them.

Also, no need for `topicFilter` variable in each test, so they each become one-liners, for example:
```
    @Test
    public void shouldNotValidateSingleLevelWildcardsCombinedWithTopicNameWithValidMultiLevelWildcard()
    {
        assertFalse(validator.isTopicFilterValid("+topic/name/#"));
    }
```